src/Processors/QueryPlan/Optimizations/tryUpdateQueryConditionCache.cpp

```diff
@@ -0,0 +1,40 @@
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/FilterStep.h>
+#include <Processors/QueryPlan/ReadFromMergeTree.h>
+#include <Storages/VirtualColumnUtils.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+void tryUpdateQueryConditionCache(const QueryPlanOptimizationSettings & optimization_settings, const Stack & stack)
+{
+    if (!optimization_settings.enable_writes_to_query_condition_cache)
+        return;
+
+    const auto & frame = stack.back();
+
+    auto * read_from_merge_tree = dynamic_cast<ReadFromMergeTree *>(frame.node->step.get());
+    if (!read_from_merge_tree)
+        return;
+
+    const auto & query_info = read_from_merge_tree->getQueryInfo();
+    auto filter_dag = query_info.filter_actions_dag;
+    if (!filter_dag || query_info.isFinal())
+        return;
+
+    if (!VirtualColumnUtils::isDeterministicInScopeOfQuery(filter_dag->getOutputs().front()))
```

Comment:
probably it is not enough. e.g. `now()` is deterministic in scope of a single query, but for condition like `ts < now() - X` some marks that are filtered out right now could pass the filter after some time.

Reply:
May be, it make sense to exclude such conditions from cached entry of condition combination?
Probably, it can be complicated to do for too complex trees, but will work for most of cases, when people have something like `WHERE ts > now() - X AND (.....)`

Reply:
Added `VirtualColumnUtils::isDeterministic` method, which should solve this problem.